kernel/src/listed_log_files.rs

```diff
@@ -41,31 +41,84 @@ pub(crate) struct ListedLogFiles {
 }
 
 /// Returns a fallible iterator of [`ParsedLogPath`] that are between the provided `start_version`
-/// (inclusive) and `end_version` (inclusive). [`ParsedLogPath`] may be a commit or a checkpoint.
-/// If `start_version` is not specified, the files will begin from version number 0. If
-/// `end_version` is not specified, files up to the most recent version will be included.
+/// (inclusive) and `end_version` (inclusive) taking into account the `log_tail` which was
+/// (ostentibly) returned from the catalog. Note that the `log_tail` must strictly adhere to being
+/// a 'tail' - that is, it goes from X..Y (inclusive) and Y must be the latest version of the
+/// table. Note that if it overlaps with commits listed from the filesystem, the `log_tail` will
+/// take precedence.
 ///
-/// Note: this calls [`StorageHandler::list_from`] to get the list of log files.
+/// [`ParsedLogPath`] may be a commit or a checkpoint. If `start_version` is not specified, the
+/// files will begin from version number 0. If `end_version` is not specified, files up to the most
+/// recent version will be included.
+///
+/// Note: this may call [`StorageHandler::list_from`] to get the list of log files unless the
+/// provided log_tail covers the entire requested range.
+///
+/// TODO: consider renaming? `maybe_list_log_files`? `list_log_files_with_tail`?
 fn list_log_files(
     storage: &dyn StorageHandler,
     log_root: &Url,
+    log_tail: Vec<ParsedLogPath>,
     start_version: impl Into<Option<Version>>,
     end_version: impl Into<Option<Version>>,
 ) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    #[cfg(debug_assertions)]
+    {
+        // log_tail is only commits
+        for entry in &log_tail {
+            debug_assert!(
+                matches!(entry.file_type, LogPathFileType::Commit),
+                "log_tail should only contain commits, found {:?}",
+                entry.file_type
+            );
+        }
+        // log_tail has no gaps nor duplicates
+        if !log_tail.is_empty() {
+            for i in 1..log_tail.len() {
+                debug_assert_eq!(
+                    log_tail[i].version,
+                    log_tail[i - 1].version + 1,
+                    "log_tail should be contiguous without gaps or duplicates"
+                );
+            }
+        }
```

Comment:
elsewhere we used the window iterator for this, in case we want to move it all into the assert statement.